Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m

```diff
@@ -222,71 +95,54 @@ - (void)commontInit
     [self.contentView addSubview:self.dateHeaderLabel];
 
-    self.textBubbleImageView = [UIImageView new];
-    self.textBubbleImageView.layoutMargins = UIEdgeInsetsZero;
-    // Enable userInteractionEnabled so that links in textView work.
-    self.textBubbleImageView.userInteractionEnabled = YES;
-
-    [self.payloadView addSubview:self.textBubbleImageView];
-
-    self.textView = [OWSMessageTextView new];
-    self.textView.backgroundColor = [UIColor clearColor];
-    self.textView.opaque = NO;
-    self.textView.editable = NO;
-    self.textView.selectable = YES;
-    self.textView.textContainerInset = UIEdgeInsetsZero;
-    self.textView.contentInset = UIEdgeInsetsZero;
-    self.textView.scrollEnabled = NO;
-
-    [self.textBubbleImageView addSubview:self.textView];
-
-    OWSAssert(self.textView.superview);
+    self.bodyTextViewCached = [self newTextView];
 
     self.footerLabel = [UILabel new];
     self.footerLabel.font = [UIFont ows_regularFontWithSize:12.f];
     self.footerLabel.textColor = [UIColor lightGrayColor];
     [self.footerView addSubview:self.footerLabel];
 
     // Hide these views by default.
-    self.textBubbleImageView.hidden = YES;
-    self.textView.hidden = YES;
+    self.bodyTextViewCached.hidden = YES;
     self.dateHeaderLabel.hidden = YES;
     self.footerLabel.hidden = YES;
 
-    [self.payloadView autoPinEdge:ALEdgeTop toEdge:ALEdgeBottom ofView:self.dateHeaderLabel];
-    [self.footerView autoPinEdge:ALEdgeTop toEdge:ALEdgeBottom ofView:self.payloadView];
-
-    [self.mediaMaskingView autoPinEdgeToSuperviewEdge:ALEdgeTop];
-
-    [self.textBubbleImageView autoPinEdge:ALEdgeTop toEdge:ALEdgeBottom ofView:self.mediaMaskingView];
-    [self.textBubbleImageView autoPinEdgeToSuperviewEdge:ALEdgeBottom];
+    [self.bubbleView autoPinEdge:ALEdgeTop toEdge:ALEdgeBottom ofView:self.dateHeaderLabel];
+    [self.footerView autoPinEdge:ALEdgeTop toEdge:ALEdgeBottom ofView:self.bubbleView];
 
     [self.footerView autoPinEdgeToSuperviewEdge:ALEdgeBottom];
     [self.footerView autoPinWidthToSuperview];
 
-    UITapGestureRecognizer *mediaTap =
-        [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(handleMediaTapGesture:)];
-    [self.mediaMaskingView addGestureRecognizer:mediaTap];
-
-    UITapGestureRecognizer *textTap =
-        [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(handleTextTapGesture:)];
-    [self.textBubbleImageView addGestureRecognizer:textTap];
+    self.contentView.userInteractionEnabled = YES;
 
-    UILongPressGestureRecognizer *mediaLongPress =
-        [[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(handleMediaLongPressGesture:)];
-    [self.mediaMaskingView addGestureRecognizer:mediaLongPress];
+    UITapGestureRecognizer *tap =
+        [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(handleTapGesture:)];
+    [self.contentView addGestureRecognizer:tap];
```

Comment:
I consolidated these gestures and changed which view they hang on, since after quoted replies we'll soon need to distinguish between taps/long presses in many regions of the message.